Migrate the settings page to Tailwind UI and Blade UI Kit

joedixon opened the pull request:
I looked through the settings pages of Tailwind UI and  like the layout I have opted for. This did mean merging the profile and the password settings onto one page (a separate form and panel for each).

<img width="1392" alt="Screenshot 2020-12-10 at 15 01 00" src="https://user-images.githubusercontent.com/3438564/101789436-38413f80-3af9-11eb-8295-08ed1579351b.png">

<img width="373" alt="Screenshot 2020-12-10 at 15 03 05" src="https://user-images.githubusercontent.com/3438564/101789452-3b3c3000-3af9-11eb-9ab0-db391eb017ab.png">

<img width="373" alt="Screenshot 2020-12-10 at 15 02 54" src="https://user-images.githubusercontent.com/3438564/101789455-3bd4c680-3af9-11eb-8a9a-8956a33dd542.png">

<img width="759" alt="Screenshot 2020-12-10 at 15 02 00" src="https://user-images.githubusercontent.com/3438564/101789456-3c6d5d00-3af9-11eb-8cbe-81bc926d591f.png">

<img width="759" alt="Screenshot 2020-12-10 at 15 01 54" src="https://user-images.githubusercontent.com/3438564/101789459-3c6d5d00-3af9-11eb-9843-9d07cb3b4914.png">

<img width="1392" alt="Screenshot 2020-12-10 at 15 01 08" src="https://user-images.githubusercontent.com/3438564/101789462-3d05f380-3af9-11eb-95f4-333bf028c95e.png">

<img width="937" alt="Screenshot 2020-12-10 at 15 03 46" src="https://user-images.githubusercontent.com/3438564/101789446-39726c80-3af9-11eb-838c-92e6a597b6bb.png">

Also resolves #576

driesvints commented:
This looks super good! But I'm thinking we can ditch the menu on the left entirely because it's already present in the navigation dropdown. I don't feel it serves much purpose there.

Sort of like this screen here: https://tailwindui.com/components/application-ui/page-examples/settings-screens#component-c29fe3f9f780c97a58f08453fc6bbec8

joedixon commented:
> This looks super good! But I'm thinking we can ditch the menu on the left entirely because it's already present in the navigation dropdown. I don't feel it serves much purpose there.
> 
> Sort of like this screen here: https://tailwindui.com/components/application-ui/page-examples/settings-screens#component-c29fe3f9f780c97a58f08453fc6bbec8

My thinking was that as we add more user features, the navbar will get a bit bloated. I liked the idea of having a single navbar item that takes you do the user area and you can nvaigate your way from there.

I'm happyt to be talked around though 😄

driesvints commented:
It just feels really out of place to me. Especially since the menu will disappear if you click any item except the settings. The navbar menu just has a handful of items and isn't bloated at all I think. Now that we have a single page for all our user settings I think it makes sense to just have the page centered without a menu next to it. If we have lots of extra settings in the future we can revise of course.

<img width="355" alt="Screenshot 2020-12-11 at 15 41 28" src="https://user-images.githubusercontent.com/594614/101916665-8b81c380-3bc7-11eb-8cde-013099ae6e3c.png">

joedixon commented:
> It just feels really out of place to me. Especially since the menu will disappear if you click any item except the settings. The navbar menu just has a handful of items and isn't bloated at all I think. Now that we have a single page for all our user settings I think it makes sense to just have the page centered without a menu next to it. If we have lots of extra settings in the future we can revise of course.

My plan was to make the sidebar persistent across all user pages so in the long-run, it wouldn't disappear. I'm happy to park that idea for now and go with the single centered column. Not sure when I'll be able to get around to it, but will try and find some time next week.

driesvints commented:
How would that look for the profile page though? What happens for the profile page when you're not logged in?

I think it's best to remove it for now and revisit it when we actually need it.

joedixon commented:
![Settings screenshot](https://user-images.githubusercontent.com/3438564/103225240-69a37300-4921-11eb-9e47-c496a3f6002c.png)

I removed the sidebar and made the panels centered on the page. I also added the ability to have a prefix icon in the form input.

driesvints commented:
@joedixon cool. Tests are failing it seems.

joedixon commented:
> @joedixon cool. Tests are failing it seems.

That will teach me to change the button text. Should be good now.

driesvints commented on app/View/Components/DangerButton.php:
```diff
@@ -0,0 +1,20 @@
+    {
+        return view('components.buttons.danger-button');
+    }
+}
```
Does this need to be a class based component? Can just be an anonymous component imo.

driesvints commented on app/View/Components/PrimaryButton.php:
```diff
@@ -0,0 +1,20 @@
+    {
+        return view('components.buttons.primary-button');
+    }
+}
```
Does this need to be a class based component? Can just be an anonymous component imo.

driesvints commented on app/View/Components/SidebarLink.php:
```diff
@@ -0,0 +1,29 @@
+    {
+        return view('components.sidebar-link');
+    }
+}
```
Does this need to be a class based component? Can just be an anonymous component imo.

driesvints commented on config/blade-ui-kit.php:
```diff
@@ -0,0 +1,125 @@
+        'toc' => Components\Markdown\ToC::class,
+        'trix' => Components\Editors\Trix::class,
+        'unsplash' => Components\Support\Unsplash::class,
+    ],
```
Let's remove all the components here which we don't use atm. Or comment them out.

driesvints commented on resources/css/forms.css:
```diff
@@ -1,104 +1,104 @@
-.choices[data-type*='select-one'] .choices__input {
+.standard .choices[data-type*='select-one'] .choices__input {
     @apply rounded-none border-b-2 border-gray-300 !important;
 }
```
Could have just wrapped the place where the file is imported in a `.standard` class heh.

driesvints commented on resources/helpers.php:
```diff
@@ -3,13 +3,27 @@
     /**
      * Sets the menu item class for an active route.
+     *
+     * @param mixed $routes
```
Remove this param. There's no type hint which already hints at it being mixed.

driesvints commented on resources/helpers.php:
```diff
@@ -3,13 +3,27 @@
+    /**
+     * Determines if the given routes are active.
+     *
+     * @param mixed $routes
```
Remove this param. There's no type hint which already hints at it being mixed.

driesvints commented on resources/helpers.php:
```diff
@@ -23,6 +37,8 @@ function md_to_html(string $markdown): string
     /**
      * Returns the route for the replyAble.
+     *
+     * @param mixed $replyAble
```
Remove this param. There's no type hint which already hints at it being mixed.

driesvints commented on resources/views/components/inputs/input.blade.php:
```diff
@@ -0,0 +1,22 @@
+            {{ $attributes }}
+            class="{{ $class }} {{ $errors->has($name) ? $errorClasses() : '' }}"
+        />
+        @if($errors->has($name))
```
```suggestion

        @if ($errors->has($name))
```

driesvints commented on resources/views/components/inputs/input.blade.php:
```diff
@@ -0,0 +1,22 @@
+            </div>
+        @endif
+    </div>
+    @if($errors->has($name))
```
```suggestion

    @if ($errors->has($name))
```

driesvints commented on resources/views/users/settings/profile.blade.php:
```diff
@@ -1,80 +1,79 @@
+                    <div class="col-span-12">
+                        <x-label for="email" />
+                        <x-input type="email" name="email" value="{{ Auth::user()->emailAddress() }}" required />
+                        @unless(Auth::user()->hasVerifiedEmail())
```
```suggestion

                        @unless(Auth::user()->hasVerifiedEmail())
```

driesvints commented on resources/views/users/settings/remove.blade.php:
```diff
@@ -0,0 +1,27 @@
+            <div class="bg-white py-6 px-4 space-y-6 sm:p-6">
+                <div>
+                    <h2 id="remove_account_heading" class="text-lg leading-6 font-medium text-red-500 uppercase">Danger Zone</h2>
+                    <p class="mt-1 text-sm leading-5 text-gray-500">Please be aware that deleting your account will also remove all of your data, including your threads and replies. This cannot be undone.</p>
```
```suggestion
                    <p class="mt-1 text-sm leading-5 text-gray-500">
                        Please be aware that deleting your account will also remove all of your data, including your threads and replies. This cannot be undone.
                    </p>
```

driesvints commented on resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php:
```diff
@@ -0,0 +1,28 @@
+                @svg($attributes->get('prefix-icon'), ['class' => 'h-5 w-5 text-gray-400'])
+            </div>
+        @endif
+        <input
```
```suggestion

        <input
```

driesvints commented on resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php:
```diff
@@ -0,0 +1,28 @@
+                'class' => 'form-input block w-full sm:text-sm sm:leading-5 mt-1' . ($attributes->get('prefix-icon') ? ' pl-10' : '') . ($errors->has($name) ? ' border-red-300 text-red-900 placeholder-red-300 focus:outline-none focus:ring-red-500 focus:border-red-500' : '')
+            ]) }}"
+        />
+        @if ($errors->has($name))
```
```suggestion

        @if ($errors->has($name))
```

driesvints commented on resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php:
```diff
@@ -0,0 +1,28 @@
+            </div>
+        @endif
+    </div>
+    @if ($errors->has($name))
```
```suggestion

    @if ($errors->has($name))
```

driesvints commented on resources/views/vendor/blade-ui-kit/components/forms/inputs/textarea.blade.php:
```diff
@@ -0,0 +1,20 @@
+            'class' => 'form-textarea block w-full transition duration-150 ease-in-out sm:text-sm sm:leading-5' . ($errors->has($name) ? ' border-red-300 text-red-900 placeholder-red-300 focus:outline-none focus:ring-red-500 focus:border-red-500' : '')
+        ]) }}"
+    >{{ old($name, $slot) }}</textarea>
+    @if($errors->has($name))
```
```suggestion

    @if($errors->has($name))
```

driesvints commented on resources/views/vendor/blade-ui-kit/components/forms/inputs/textarea.blade.php:
```diff
@@ -0,0 +1,20 @@
+            <x-heroicon-o-exclamation-circle class="h-5 w-5 text-red-500" />
+        </div>
+    @endif
+    @if($errors->has($name))
```
```suggestion

    @if($errors->has($name))
```

driesvints reviewed (changes requested):
Nice job. I left some remarks.

A few notes in general:
- Always try to leave some breathing room before or after an if directive in Blade views when there's elements on the same level before or after it.
- If directives always have a space between `if` and the opening `(`
- I see a lot of `{{` syntax being used in Blade component value attributes. Did you test all of these? Because afaik this isn't supported on Blade components. You should try to use the `:value="Auth::user()->email"` syntax instead.

driesvints commented:
@joedixon tests seem to be failing.

joedixon commented:
> @joedixon tests seem to be failing.

Fixed!
